controllers/adapters_controller.js

```diff
@@ -0,0 +1,195 @@
+  adapterManager.cancelAddSomeDevice();
+  response.status(204).send();
+});
+
+/**
+ * Cancel Action: remove a device from the adapter.
+ */
+adaptersController.delete('/removeSomeDevice', (request, response) => {
+  adapterManager.cancelRemoveSomeDevice();
+  response.status(204).send();
+});
+
+/**
+ * Get a particular adapter.
+ */
+adaptersController.get('/:adapterId/', (request, response) => {
+  var adapterId = request.params['adapterId'];
+
+  if (adapterId === 'addSomeDevice') {
+    // Action: add a new device to the adapter.
+    adapterManager.addSomeDevice().then((device) => {
+      console.log('addSomeDevice added id', device.getId(), 'name', device.getName());
+    }, (str) => {
+      console.log('addSomeDevice cancelled');
+    });
+    response.json({
+      'name': 'addSomeDevice',
+      'href': request.originalUrl,
+    });
+    return
+  }
+
+  if (adapterId === 'removeSomeDevice') {
+    // Action: remove a device from the adapter.
+    adapterManager.removeSomeDevice().then((device) => {
+      console.log('removeSomeDevice removed id', device.getId(), 'name', device.getName());
+    }, (str) => {
+      console.log('removeSomeDevice cancelled');
+    });
+    response.json({
+      'name': 'removeSomeDevice',
+      'href': request.originalUrl,
+    });
+    return;
+  }
+
+  var adapter = adapterManager.getAdapter(adapterId);
+  if (adapter) {
+    response.json(adapter.asDict());
+    return;
+  }
+  response.status(404).send('Adapter "' + adapterId + '" not found.');
+});
+
+/**
+ * Get a list of devices associated with a particular adapter.
+ */
+adaptersController.get('/:adapterId/devices/', (request, response) => {
+  var adapterId = request.params['adapterId'];
+  var adapter = adapterManager.getAdapter(adapterId);
+  if (adapter) {
+    adapter.discoverDevices().then(function(devices) {
+      var deviceList = [];
+      for (var deviceId in adapter.devices) {
+        var device = adapter.devices[deviceId];
+        deviceList.push(device.asDict());
+      }
+      response.json(deviceList);
+    });
+    return;
+  }
+  response.status(404).send('Adapter "' + adapterId + '" not found.');
+});
+
+/**
+ * Get a particular device associated with a particular adapter.
+ */
+adaptersController.get('/:adapterId/devices/:deviceId/', (request, response) => {
+  var adapterId = request.params['adapterId'];
+  var deviceId = request.params['deviceId'];
+  var adapter = adapterManager.getAdapter(adapterId);
+  if (adapter) {
+    var device = adapter.getDevice(deviceId);
+    if (device) {
+      response.json(device.asDict());
+      return;
+    }
+    response.status(404).send('Adapter "' + adapterId + '" device "' + deviceId + '" not found.');
+    return;
+  }
+  response.status(404).send('Adapter "' + adapterId + '" not found.');
+});
+
+/**
+ * Get a particular device associated with a particular adapter.
+ */
+adaptersController.get('/:adapterId/devices/:deviceId/:attributeId', (request, response) => {
```

Comment:
I think ThingController will also need to call methods on AdapterManager in order to get and set properties on devices (not just interact with the database), and we'll need a controller that can start and stop pairing and get a list of adapters. I assumed that was what this code was for.

But if you really think we need a REST API for integration testing then yes, maybe it should be called test/ or similar.